File: assets/templates/check-token.js
// Token Safety Checker for Pharos
// Usage: node check-token.js <token_address> [rpc_url]

const { ethers } = require("ethers");

const ERC20_ABI = [
  "function name() view returns (string)",
  "function symbol() view returns (string)",
  "function totalSupply() view returns (uint256)",
  "function decimals() view returns (uint8)",
  "function balanceOf(address) view returns (uint256)",
  "function owner() view returns (address)",
];

async function checkTokenSafety(tokenAddress, rpcUrl) {
  const provider = new ethers.JsonRpcProvider(rpcUrl);
  const token = new ethers.Contract(tokenAddress, ERC20_ABI, provider);
  
  const report = {
    address: tokenAddress,
    chain: "Pharos",
    timestamp: new Date().toISOString(),
    findings: [],
    score: 100,
  };

  try {
    report.name = await token.name();
    report.symbol = await token.symbol();
    report.totalSupply = (await token.totalSupply()).toString();
    report.decimals = await token.decimals();
  } catch (e) {
    report.findings.push({ severity: "CRITICAL", finding: "Not a valid ERC-20 token" });
    report.score = 0;
    return report;
  }

  // Check owner
  try {
    const owner = await token.owner();
    if (owner !== ethers.ZeroAddress) {
      report.findings.push({ severity: "MEDIUM", finding: `Has owner: ${owner}` });
      report.score -= 15;
    }
  } catch {
    report.findings.push({ severity: "LOW", finding: "No owner function (decentralized)" });
  }

  // Check if mintable
  try {
    await token.mint.staticCall(ethers.ZeroAddress, 0);
    report.findings.push({ severity: "HIGH", finding: "Token is MINTABLE" });
    report.score -= 20;
  } catch {
    report.findings.push({ severity: "LOW", finding: "Not mintable" });
  }

  report.score = Math.max(0, report.score);
  report.verdict = report.score >= 70 ? "SAFE" : report.score >= 50 ? "CAUTION" : "AVOID";
  
  return report;
}

// CLI
if (require.main === module) {
  const token = process.argv[2];
  const rpc = process.argv[3] || "https://atlantic.dplabs-internal.com";
  if (!token) { console.log("Usage: node check-token.js <token_address> [rpc_url]"); process.exit(1); }
  checkTokenSafety(token, rpc).then(r => console.log(JSON.stringify(r, null, 2)));
}

module.exports = { checkTokenSafety };

File: assets/templates/check-token.py
# Token Safety Checker for Pharos (Python)
# Usage: python check-token.py <token_address> [rpc_url]

import json
import sys
from web3 import Web3

ERC20_ABI = [
    {"constant": True, "inputs": [], "name": "name", "outputs": [{"name": "", "type": "string"}], "type": "function"},
    {"constant": True, "inputs": [], "name": "symbol", "outputs": [{"name": "", "type": "string"}], "type": "function"},
    {"constant": True, "inputs": [], "name": "totalSupply", "outputs": [{"name": "", "type": "uint256"}], "type": "function"},
    {"constant": True, "inputs": [], "name": "decimals", "outputs": [{"name": "", "type": "uint8"}], "type": "function"},
    {"constant": True, "inputs": [{"name": "", "type": "address"}], "name": "balanceOf", "outputs": [{"name": "", "type": "uint256"}], "type": "function"},
]

def check_token_safety(token_address: str, rpc_url: str) -> dict:
    w3 = Web3(Web3.HTTPProvider(rpc_url))
    
    report = {
        "address": token_address,
        "chain": "Pharos",
        "timestamp": __import__("datetime").datetime.utcnow().isoformat(),
        "findings": [],
        "score": 100,
    }

    try:
        token = w3.eth.contract(address=Web3.to_checksum_address(token_address), abi=ERC20_ABI)
        report["name"] = token.functions.name().call()
        report["symbol"] = token.functions.symbol().call()
        report["totalSupply"] = str(token.functions.totalSupply().call())
        report["decimals"] = token.functions.decimals().call()
    except Exception as e:
        report["findings"].append({"severity": "CRITICAL", "finding": f"Not a valid ERC-20: {e}"})
        report["score"] = 0
        return report

    # Check if contract has code (not a simple EOA)
    code = w3.eth.get_code(Web3.to_checksum_address(token_address))
    if code == b"" or code == b"0x":
        report["findings"].append({"severity": "CRITICAL", "finding": "No contract code at address"})
        report["score"] = 0

    report["score"] = max(0, report["score"])
    report["verdict"] = "SAFE" if report["score"] >= 70 else "CAUTION" if report["score"] >= 50 else "AVOID"
    
    return report


if __name__ == "__main__":
    token = sys.argv[1] if len(sys.argv) > 1 else None
    rpc = sys.argv[2] if len(sys.argv) > 2 else "https://atlantic.dplabs-internal.com"
    
    if not token:
        print("Usage: python check-token.py <token_address> [rpc_url]")
        sys.exit(1)
    
    result = check_token_safety(token, rpc)
    print(json.dumps(result, indent=2))
